proposals/attach-to-property.md

```diff
@@ -0,0 +1,142 @@
+# Attach to delegate property object
+
+* **Type**: Design proposal
+* **Author**: Ilya Ryzhenkov
+* **Contributors**:
+* **Status**:
+* **Prototype**: Was implemented and then disabled
+
+## Summary
+
+Recognize special operator function available for delegate object and call it right after it has been stored
+into a field for delegation.
+
+## Glossary
+
+- "delegate instance": instance on which property's get/set accessors execute respective getValue/setValue calls.
+- "host instance": instance containing delegated properties as members, or being extended with extension property.
+- "attachTo": placeholder for the name of the operator function
+- "constructor" means either primary constructor, or each of secondary constructor if there is no primary constructor
+
+## Motivation / use cases
+
+Make delegate instance aware of host instance and property it was used with.
+
+- Register host instance in external table of managed entities, e.g. database schema, component container.
+- Register delegate properties in host instance with callback for "metadata", e.g. for database columns.
+- Call back into host instance with additional information, e.g. property metadata, delegate instance.
+  Can be used to communicate delegate instance back to the host instance for later use, by declaring member
+  extension `attachTo` function with receiver of delegate type.
+- Execute custom code before first access to property, e.g. `async` property that is being calculated
+  in the background, and blocks as needed when trying to use the value. With coroutines, we can suspend instead
+  of blocking.
+
+## Description
+
+Client code:
+```
+class HostType {
+   val name by DelegateType()
+}
+```
+
+Attach operator:
+```
+operator fun DelegateType.attachTo(instance: HostType, property: KProperty<*>) : Unit`
+```
+
+- No extra instances are created, all information is already available
+
+## Alternative names
+
+- bindToInstance
+- delegatedFrom
+
+## Implementation
+
+### Syntax
+
+No new syntax is required.
+
+### Resolution
+
+- resolving the `attachTo` function happens similar to resolving `getValue`/`setValue` functions.
+- if there are more than one `attachTo` functions available, overload resolution selects best candidate.
```

Comment:
wrt to type of `instance` being the _default type_ of the containing class.

What if the property is a not a member? We must say explicitly that `null` as passed for the first parameter